include/range/v3/utility/counted_iterator.hpp

```diff
@@ -25,183 +25,344 @@ namespace ranges
-                // counted iterators
-                friend I uncounted(counted_iterator<I, D> i)
+                CONCEPT_ASSERT(Iterator<I>());
+                template<typename, typename> friend struct counted_iterator;
+                I current_{};
+                difference_type_t<I> cnt_{0};
+            public:
+                using iterator_type = I;
+                using difference_type = difference_type_t<I>;
+
+                counted_iterator() = default;
+
+                counted_iterator(I x, difference_type_t<I> n)
+                  : current_(std::move(x)), cnt_(n)
                 {
-                    return i.base();
+                    RANGES_ENSURE(n >= 0);
                 }
-                friend counted_iterator<I, D>
-                recounted(counted_iterator<I, D> const &j, I i, iterator_difference_t<I> n)
+
+                template<typename I2,
+                    CONCEPT_REQUIRES_(ConvertibleTo<I2, I>())>
+                counted_iterator(const counted_iterator<I2>& i)
+                  : current_(i.current_), cnt_(i.cnt_)
+                {}
+
+                template<typename I2,
+                    CONCEPT_REQUIRES_(ConvertibleTo<I2, I>())>
+                counted_iterator& operator=(const counted_iterator<I2>& i)
                 {
-                    RANGES_ASSERT(!ForwardIterator<I>() || ranges::next(j.base(), n) == i);
-                    return {i, j.count() - n};
+                    current_ = i.current_;
+                    cnt_ = i.cnt_;
                 }
-                template<typename II = I,
-                    CONCEPT_REQUIRES_(Readable<II>())>
-                iterator_rvalue_reference_t<II> move() const
-                    noexcept(noexcept(iter_move(std::declval<II const &>())))
+
+                I base() const
                 {
-                    return iter_move(it_);
+                    return current_;
                 }
-                CONCEPT_REQUIRES(Readable<I>())
-                iterator_reference_t<I> read() const
+
+                difference_type_t<I> count() const
                 {
-                    return *it_;
+                    return cnt_;
                 }
-                template<typename T,
-                    CONCEPT_REQUIRES_(ExclusivelyWritable_<I const, T &&>())>
-                void write(T && t) const
+
+                reference_t<I> operator*()
+                RANGES_AUTO_RETURN_NOEXCEPT
+                (
+                    *current_
+                )
+                template<typename I2 = I, CONCEPT_REQUIRES_(Readable<I2 const>())>
+                reference_t<I> operator*() const
+                RANGES_AUTO_RETURN_NOEXCEPT
+                (
+                    *static_cast<I2 const &>(current_)
+                )
+
+                counted_iterator& operator++()
                 {
-                    *it_ = (T &&) t;
+                    RANGES_ENSURE(cnt_ > 0);
```

Comment:
It's side-effect free and zero-cost, I'd use `RANGES_EXPECT` here and try to help the optimizer. `RANGES_ASSERT` is now really only used for side-effect-free but potentially expensive checks.